File: src/main/java/cn/qianlicao/initialization/Task.java
package cn.qianlicao.initialization;

import java.util.HashMap;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Created by dongyayun on 16/2/18.
 */
public abstract class Task implements Runnable {
    private final AtomicBoolean mIsCancelled = new AtomicBoolean();
    private HashMap<String, Task> mSons = new HashMap();
    private int mTaskStatus = Status.STATUS_INIT;

    private Task mFather;

    private String mTaskName;

    public Task(String name) {

    }


    @Override
    public void run() {
        if (isCancelled())
            return;


        if (mFather != null) {
            Initialization.getInstance().getThreadPool().submit(mFather);
            setTaskStatus(Status.STATUS_WAITING);
            try {
                wait();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }


        if (isCancelled())
            return;

        onDoSomething();

        onFinish();


        for (Task t : mSons.values()) {
            if (t.getTaskStatus() == Status.STATUS_INIT) {
                Initialization.getInstance().getThreadPool().submit(t);
            } else if (t.getTaskStatus() == Status.STATUS_WAITING) {
                t.notify();
            } else {
                //do nothing
            }
        }
    }

    protected void onDoSomething() {
        setTaskStatus(Status.STATUS_DOING);
    }

    protected void onFinish() {
        setTaskStatus(Status.STATUS_DONE);
    }

    public void cancel() {
        if (mFather != null) {
            mFather.cancel();
            mFather = null;
        }

        for (Task task : mSons.values()) {
            task.cancel();
        }

        mSons.clear();

        mIsCancelled.set(true);
    }

    private boolean isCancelled() {
        return mIsCancelled.get();
    }


    public int getTaskStatus() {
        return mTaskStatus;
    }

    public void setTaskStatus(int taskStatus) {
        mTaskStatus = taskStatus;
    }

    public Task getFather() {
        return mFather;
    }

    public void setFather(Task father) {
        mFather = father;
        mFather.addSon(this);
    }

    public HashMap<String, Task> getSons() {
        return mSons;
    }

    public void setSons(HashMap<String, Task> sons) {
        mSons = sons;
    }

    public void addSon(Task task) {
        mSons.put(task.getTaskName(), task);
    }


    public void removeSon(Task task) {
        String name = task.getTaskName();


        Task task1 = mSons.get(name);
        if (task1 == null)
            return;

        task1.cancel();
        mSons.remove(name);
    }

    public String getTaskName() {
        return mTaskName;
    }

    public void setTaskName(String taskName) {
        mTaskName = taskName;
    }
}

File: src/main/java/cn/qianlicao/initialization/Task.kt
package cn.qianlicao.initialization

import android.util.Log
import java.util.*
import java.util.concurrent.CountDownLatch
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Created by dongyayun on 16/2/18.
 */
abstract class Task(name: String) : Runnable {
    private val mIsCancelled = AtomicBoolean()
    var sons: HashMap<String, Task> = HashMap()
    var taskStatus = Status.STATUS_INIT

    var countdownLatch: CountDownLatch = CountDownLatch(1);

    var father: Task? = null
        set(value) {

            //father =value   千万不能这样用,否则会无限循环set
            field = value
            field!!.addSon(this)
        }

    var taskName: String = name


    override fun run() {
        if (isCancelled)
            return

        if (father != null) {

            Log.d("Task", taskName + "has father")

            Initialization.instance.threadPool.submit(father)
            taskStatus = Status.STATUS_WAITING
            countdownLatch.await()
        }


        if (isCancelled)
            return

        onDoSomething()

        onFinish()


        for (t in sons.values) {
            if (t.taskStatus == Status.STATUS_INIT) {
                Initialization.instance.threadPool.submit(t)
            } else if (t.taskStatus == Status.STATUS_WAITING) {
                t.countdownLatch.countDown();
            } else {
                //do nothing
            }
        }
    }

    protected open fun onDoSomething() {
        if (father != null) {
            Log.d("Task", taskName + "continue doing something")
        } else {
            Log.d("Task", taskName + "do something")
        }


        taskStatus = Status.STATUS_DOING
    }

    protected fun onFinish() {
        taskStatus = Status.STATUS_DONE
    }

    fun cancel() {
        if (father != null) {
            father!!.cancel()
            father = null
        }

        for (task in sons.values) {
            task.cancel()
        }

        sons.clear()

        mIsCancelled.set(true)
    }

    private val isCancelled: Boolean
        get() = mIsCancelled.get()

    fun addSon(task: Task) {
        sons.put(task.taskName, task)
    }


    fun removeSon(task: Task) {
        val name = task.taskName


        val task1 = sons[name] ?: return

        task1.cancel()
        sons.remove(name)
    }

}
